### PR TITLE
Optionally publish acceleration state

### PR DESCRIPTION
**Pull request by reinzor**
Expose acceleration state via a geometry_msgs/AccelWithCovarianceStamped message. This is configurable by the `publish_acceleration` parameters which defaults to `false`.

#314

### CODE REVIEW

File: include/robot_localization/ros_filter.h [hunk 1]
```diff
@@ -165,10 +166,11 @@ template<class T> class RosFilter
                    std::vector<int> &updateVector);
 
     //! @brief Retrieves the EKF's output for broadcasting
-    //! @param[out] message - The standard ROS odometry message to be filled
+    //! @param[out] odomMsg - The standard ROS odometry message to be filled
+    //! @param[out] accelMsg - The standard ROS acceleration message to be filled
     //! @return true if the filter is initialized, false otherwise
     //!
-    bool getFilteredOdometryMessage(nav_msgs::Odometry &message);
+    bool getFilteredOdometryAndAccelMessage(nav_msgs::Odometry &odomMsg, geometry_msgs::AccelWithCovarianceStamped& accelMsg);
```
Comment (ayrton04) [inline include/robot_localization/ros_filter.h]: I think I'd prefer that we add a second method for assembling the acceleration message. Then we can fill it conditionally base on the value of `publishAcceleration_`, and we're not changing this method signature.

File: src/ros_filter.cpp [hunk 2]
```diff
@@ -268,27 +269,27 @@ namespace RobotLocalization
...
-      message.pose.pose.orientation.y = quat.y();
-      message.pose.pose.orientation.z = quat.z();
-      message.pose.pose.orientation.w = quat.w();
-      message.twist.twist.linear.x = state(StateMemberVx);
-      message.twist.twist.linear.y = state(StateMemberVy);
-      message.twist.twist.linear.z = state(StateMemberVz);
-      message.twist.twist.angular.x = state(StateMemberVroll);
-      message.twist.twist.angular.y = state(StateMemberVpitch);
-      message.twist.twist.angular.z = state(StateMemberVyaw);
+      //! Fill out the odometry message
```
Comment (ayrton04) [inline src/ros_filter.cpp]: To keep the PR focused, I'd prefer that we not change anything unless needed.

File: src/ros_filter.cpp [hunk 3]
```diff
@@ -646,9 +668,12 @@ namespace RobotLocalization
     FilterUtilities::appendPrefix(tfPrefix, baseLinkFrameId_);
     FilterUtilities::appendPrefix(tfPrefix, worldFrameId_);
 
-    // Whether we're publshing the world_frame->base_link_frame transform
+    // Whether we're publishing the world_frame->base_link_frame transform
```
Comment (ayrton04) [inline src/ros_filter.cpp]: I realize this is a typo, but I'd rather have a PR that addresses comments and such specifically.

File: src/ros_filter.cpp [hunk 4]
```diff
@@ -1630,8 +1655,11 @@ namespace RobotLocalization
     worldBaseLinkTransMsg_.transform = tf2::toMsg(tf2::Transform::getIdentity());
     mapOdomTransMsg.transform = tf2::toMsg(tf2::Transform::getIdentity());
 
-    // Publisher
-    ros::Publisher positionPub = nh_.advertise<nav_msgs::Odometry>("odometry/filtered", 20);
+    // Publishers
+    ros::Publisher odomPub = nh_.advertise<nav_msgs::Odometry>("odometry/filtered", 20);
+    ros::Publisher accelPub;
+    if (publishAcceleration_)
```
Comment (ayrton04) [inline src/ros_filter.cpp:1678]: Please use curly braces, even for one-line conditional statements.

File: src/ros_filter.cpp [hunk 5]
```diff
@@ -1647,28 +1675,29 @@ namespace RobotLocalization
       integrateMeasurements(curTime);
 
       // Get latest state and publish it
-      nav_msgs::Odometry filteredPosition;
+      nav_msgs::Odometry odomMsg;
```
Comment (ayrton04) [inline src/ros_filter.cpp]: As above: if we don't have to change this, I'd prefer that we leave it alone, even if the variable name is admittedly not great. This rename ends up creating a lot of changed lines in the PR that don't actually change any functionality.

**Review by ayrton04 — CHANGES_REQUESTED** [review]
Looks like a nice addition, thank you. I'd ask that we keep the PR focused on only the changes that need to be made; I give some more detail in the individual comments.

Thank you for the PR!

**Comment by reinzor** [comment]
I just processed the comments and committed to make the changes minimal

**Comment by ayrton04** [comment]
Looks good. I will test this out and merge. Thanks again!

**Comment by ayrton04** [comment]
Just testing PR jobs on the build server. Please ignore the close and open.

**Comment by ayrton04** [comment]
Testing PR job again.

**Comment by ayrton04** [comment]
Ignore build failure. That's a separate issue with the CMakeLists file. I pulled this down and tested it, and it looks good. Thanks again for the PR!

**Pull request merged**